packages/protocol/contracts/ACE/ACE.sol

```diff
@@ -186,6 +187,148 @@ contract ACE is IAZTEC {
+    *      
+    * @param _proof the AZTEC proof object
+    * @param _proofData the mint proof construction data
+    * @param _proofSender the Ethereum address of the original transaction sender. It is explicitly assumed that
+    *        an asset using ACE supplies this field correctly - if they don't their asset is vulnerable to front-running
+    * Unnamed param is the AZTEC zero-knowledge proof data
+    * @return two `bytes` objects. The first contains the new confidentialTotalSupply note and the second contains the
+    * notes that were created. Returned so that a zkAsset can emit the appropriate events
+    */
+    function mint(
+        uint24 _proof,
+        bytes calldata _proofData,
+        address _proofSender
+    ) external returns (bytes memory, bytes memory) {
+        
+        NoteRegistry storage registry = registries[msg.sender];
+        require(registry.flags.active == true, "note registry does not exist for the given address");
+        require(registry.flags.canAdjustSupply == true, "this asset is not mintable");
+        
+        // Check that it's a mintable proof
+        (, uint8 category, ) = _proof.getProofComponents();
+
+        require(category == uint8(ProofCategory.MINT), "this is not a mint proof");
+
+        bytes memory _proofOutputs = this.validateProof(_proof, _proofSender, _proofData);
+        require(_proofOutputs.getLength() > 0, "call to validateProof failed");
+
+        // Dealing with notes representing totals
+        (bytes memory oldTotal,  // inputNotesTotal
+        bytes memory newTotal, // outputNotesTotal
+        ,
+        int256 publicValueTotal) = _proofOutputs.get(0).extractProofOutput();
+
+        require(publicValueTotal == 0, "mint transactions cannot have a public value");
+        require(newTotal.getLength() == 1, "new total minted must be represented by one note");
+        require(oldTotal.getLength() == 1, "old total minted must be represented by one note");
+    
+        // Check the previous confidentialTotalSupply, and then assign the new one
+        (, bytes32 oldTotalNoteHash, ) = oldTotal.get(0).extractNote();        
+
+        require(oldTotalNoteHash == registry.confidentialTotalMinted, "provided total supply note does not match");
+        (, bytes32 newTotalNoteHash, ) = newTotal.get(0).extractNote();
+        registry.confidentialTotalMinted = newTotalNoteHash;
+
+
+        // Dealing with minted notes
+        (bytes memory inputNotes, // input notes
+        bytes memory mintedNotes, // output notes
+        ,
+        int256 publicValue
+        ) = _proofOutputs.get(1).extractProofOutput();
+
+        require(publicValue == 0, "mint transactions cannot have a public value");
+        require(mintedNotes.getLength() >= 1, "must have at least one minted note");
+        require(inputNotes.getLength() == 0, "second proofOutput object can not have input notes");
+
+        updateOutputNotes(mintedNotes);
+        return(newTotal, mintedNotes);
```

Comment:
I think we should return the outputs of 'mint' and 'burn' proofs via the `bytes proofOutputs` format in the spec. The more generic our interface the better - we don't know what future data we might want to add to the outputs of new mint and burn proofs.

Reply:
Good idea, done